torch/csrc/jit/source_range.h

```diff
@@ -9,25 +9,65 @@
 namespace torch {
 namespace jit {
 
+struct SourceRange;
+using SourceRangeRecord = std::tuple<size_t, std::shared_ptr<SourceRange>>;
+using SourceRangeRecords = std::vector<SourceRangeRecord>;
+
+// Class that keeps track of a serialized debug info table and lazily
+// unpacks it on query.
+class CAFFE2_API DebugInfo {
+ public:
+  // I would use a tuple here but it seems like some C++ compilers have issues
+  // std::move'ing tuples where an element has a deleted assignment operator...
+  struct SerializedDebugInfo {
+    SerializedDebugInfo(at::DataPtr&& data, size_t size)
+        : data(std::move(data)), size(size) {}
+
+    SerializedDebugInfo(SerializedDebugInfo&& rhs)
+        : SerializedDebugInfo(std::move(rhs.data), rhs.size) {}
+
+    at::DataPtr data;
+    size_t size;
+  };
+
+  DebugInfo() = default;
+
+  DebugInfo(SerializedDebugInfo info) : info_(std::move(info)) {}
+
+  std::shared_ptr<SourceRange> query(const SourceRange& q);
+
+ private:
+  void deserialize();
+
+  SerializedDebugInfo info_;
+  c10::optional<SourceRangeRecords> deserialized_records_;
+};
+
 // Source represents a code segment. It keeps track of:
 //  - text : the text of the code segment
 //  - filename (optional) : if present, represents the name of the file from
 //                          which the code semgemnt originated.
 //  - starting_line_no : represents the line in the original file where the
 //                       code segment started.
 struct Source {
-  explicit Source(std::string text)
-      : text_(std::move(text)), filename_(c10::nullopt) {
+  explicit Source(
+      std::string text,
+      std::shared_ptr<DebugInfo> debug_info = nullptr)
```

Comment:
DebugInfo is not a great name here because it doesn't really describe what it does. After all, all of SourceRange/Source is really 'debug info'. It would be better if this were a chain. i.e. generating_source, the source that generated this source.